dockerfile_parse/parser.py

```diff
@@ -533,6 +581,85 @@ def _add_instruction(self, instruction, value):
+        :param at_start: adds at the beginning (after FROM) of the stage(s) instead of the end.
+        """
+        assert len(lines) > 0
+        all_stages = kwargs.pop('all_stages', False)
+        at_start = kwargs.pop('at_start', False)
+        assert not kwargs, "Unknown keyword argument(s): {}".format(kwargs.keys())
+
+        froms = [
+            instr for instr in self.structure
+            if instr['instruction'] == 'FROM'
+        ] or [{'endline': -1}]  # no FROM? fake one before the beginning
+        if not all_stages:  # only modify the last
+            froms = [froms[-1]]
+
+        df_lines = self.lines
+        # make sure last line has a newline if lines are to be appended
+        if df_lines and not at_start:
+            df_lines[-1] = _endline(df_lines[-1])
+
+        # iterate through the stages in reverse order
+        # so adding lines doesn't invalidate line numbers from structure dicts.
+        # first add a bogus instruction to represent EOF in our iteration.
+        froms.append({'startline': len(df_lines) + 1})
+        for stage in range(len(froms)-2, -1, -1):  # e.g. 0 for single or 2, 1, 0 for 3 stages
+            start, finish = froms[stage], froms[stage+1]
+            linenum = start['endline'] + 1 if at_start else finish['startline']
+            df_lines[linenum:linenum] = [_endline(line) for line in lines]
+
+        self.lines = df_lines
+
+    def add_lines_at(self, anchor, *lines, **kwargs):
+        """
+        Add lines at a specific location in the file.
+        :param anchor: structure_dict|line_str|line_num a reference to where adds should occur
+        :param lines: one or more lines to add to the content
+        :param replace: if True -- replace the anchor
+        :param after: if True -- insert after the anchor (conflicts with "replace")
+        """
+        assert len(lines) > 0
+        replace = kwargs.pop('replace', False)
+        after = kwargs.pop('after', False)
+        assert not (after and replace)
+        assert not kwargs, "Unknown keyword argument(s): {}".format(kwargs.keys())
+
+        # find the line number for the insertion
+        df_lines = self.lines
+        if isinstance(anchor, int):  # line number, just validate
+            assert anchor in range(len(df_lines))
+            if replace:
+                del df_lines[anchor]
+        elif isinstance(anchor, dict):  # structure
+            assert anchor in self.structure, "Current structure does not match: {}".format(anchor)
+            if replace:
+                df_lines[anchor['startline']:anchor['endline'] + 1] = []
+            anchor = anchor['startline']
+        elif isinstance(anchor, string_types):  # line contents
+            matches = [index for index, text in enumerate(df_lines) if text == anchor]
+            if not matches:
+                raise RuntimeError("Cannot find line in the build file:\n" + anchor)
+            anchor = matches[-1]
```

Comment:
Oh, the *last* match? Can we have a test case that checks this?

Reply:
Modified test_add_lines_at to cover this